klippy/extras/xpt2046.py

```diff
@@ -0,0 +1,306 @@
+        self._menu = menu
+        self._function = function
+
+    def invoke(self, eventtime):
+        self._menu.key_event(self._function, eventtime)
+
+class Xpt2046GcodeAction(object):
+    def __init__(self, printer, config, template_name):
+        gcode_macro = printer.load_object(config, 'gcode_macro')
+        self._gcode = printer.lookup_object("gcode")
+        self._template = gcode_macro.load_template(config, template_name)
+
+    def invoke(self, eventtime):
+        del eventtime
+        self._gcode.run_script(self._template.render())
+
+class Xpt2046DummyAction(object):
+    # pylint: disable=R0201
+    def invoke(self, eventtime):
+        del eventtime
+
+def _create_action(printer_object, config, action_prefix):
+    menu_function = config.get(action_prefix + '_function', None)
+    if menu_function:
+        menu_object = printer_object.lookup_object("menu")
+        return Xpt2046MenuAction(menu_object, menu_function)
+    if config.get(action_prefix + '_gcode', None):
+        return Xpt2046GcodeAction(printer_object, config,
+                                  action_prefix + '_gcode')
+    return None
+
+class Xpt2046Button(object):
+    def __init__(self, points):
+        self.points = []
+
+        max_x = None
+        points = points.split('\n')
+        for point in points:
+            if point:
+                new_point = _parse_coordinate(point)
+                if max_x is None or max_x < new_point[0]:
+                    max_x = new_point[0]
+
+                self.points.append(new_point)
+
+        if len(self.points) < 3:
+            raise XptConfigException(
+                "Polygon with at least 3 vertices expected")
+
+        self._max_ray_x = max_x + 100 # X coordinate used for ray casting
+
+    def check_xy(self, location):
+        def ccw(_a, _b, _c):
+            return ((_c[1]-_a[1]) * (_b[0]-_a[0]) >
+                    (_b[1]-_a[1]) * (_c[0]-_a[0]))
+
+        def intersect(_a, _b, _c, _d):
+            return (ccw(_a, _c, _d) != ccw(_b, _c, _d) and
+                    ccw(_a, _b, _c) != ccw(_a, _b, _d))
+
+        result = False
+        # use a 0.1 offset (while polygon points are truncated) to ensure
+        # that the casted ray can never be colinear with a polygon vertex
+        outside_location = (self._max_ray_x, location[1] + 0.1)
+        for i in range(len(self.points)-1):
+            if intersect(self.points[i], self.points[i+1],
+                         location, outside_location):
+                result = not result
+
+        if intersect(self.points[-1], self.points[0],
+                     location, outside_location):
+            result = not result
+
+        return result
+
+    # The following two methods are to be overridden by derived classes
+    # pylint: disable=R0201
+
+    # Called when a button is clicked. Can return a timestamp for an expected
+    # repeat event.
+    def clicked(self, eventtime):
+        del eventtime
```

Comment:
What is the intent of `del eventtime`?

Reply:
Making pylint happy:

http://pylint-messages.wikidot.com/messages:w0613

Reply:
Okay.  FWIW, I don't think that pylint test has value.  Might be better to just disable that lint check.